geo/src/algorithm/convex_hull/graham.rs

```diff
@@ -0,0 +1,165 @@
+use super::swap_remove_to_first;
+use crate::algorithm::kernels::*;
+use crate::{Coordinate, LineString};
+
+/// Graham Scan: https://en.wikipedia.org/wiki/Graham_scan
+/// This algorithm also provides an option to obtain all
+/// points on the convex hull as opposed to only the
+/// strictly convex points.
+pub fn graham_hull<T>(mut points: &mut [Coordinate<T>], include_on_hull: bool) -> LineString<T>
+where
+    T: HasKernel,
+{
+    if points.len() < 4 {
+        // Nothing to build with fewer than four points.
+
+        // Remove repeated points unless colinear points
+        // are to be included.
+        let mut ls: LineString<T> = if include_on_hull {
+            points.iter().copied().collect()
+        } else {
+            points
+                .iter()
+                .enumerate()
+                .filter_map(|(i, pt)| {
+                    // Do not care if first and last are
+                    // same, as we anyway close the linestring
+                    if i == 0 || pt != &points[i - 1] {
+                        Some(*pt)
+                    } else {
+                        None
+                    }
+                })
```

Comment:
is this the same?

```suggestion
                .filter(|(i, pt)| {
                    // Do not care if first and last are
                    // same, as we anyway close the linestring
                    i == 0 || pt != &points[i - 1]
                })
```

Reply:
Yeah, the logic is the same.  However, if we just filter, we'll end up with Vec<(usize, Coord)>, but we want only the second component.  

Good catch of the repeated code.  I'd abstracted it out as `trivial_hull` but forgot to change the graham impl.

Reply:
Ah yeah gotcha. Also, is this the same as using https://doc.rust-lang.org/std/vec/struct.Vec.html#method.dedup ?

Reply:
Yeah, exactly;  never knew this method!   #TIL